### ISSUE TITLE
libproc_macro: get crate for span or support $crate metavar in quote!{}

### ISSUE THREAD

**Issue by abonander** [rust-lang/rust#49797]
[One Reddit user was asking how to reference the invoking crate in a proc-macro](https://www.reddit.com/r/rust/comments/88xvgg/hey_rustaceans_got_an_easy_question_ask_here/dx0kp1f/), suggesting that `quote!{}` should support the `$crate` metavar. I don't think this is too unreasonable since `$` is already used for unquoting variables.

Alternately, I think `Span` could have a method telling us what crate it comes from as a `String`.

cc @alexcrichton @dtolnay

**Comment by Badel2** [comment]
I don't know if my use case can be resolved this way: I have a crate [enum-vec](https://github.com/Badel2/enum-vec) but when I want to use it I have to include 3 crates:

```rust
#[macro_use]
extern crate enum_like_derive;
extern crate enum_like;
extern crate enum_vec;
```

I thought that maybe if I reexport `EnumLike` from `enum_vec` (since `enum_vec` already depends on `enum_like`) I could do `use enum_vec::EnumLike;` instead of including the `enum_like` crate, saving one precious line of code. But my proc macro `enum_like_derive` is hardcoded to use `::enum_like::EnumLike`, so I was wondering which is the correct way to refer to `EnumLike`.

I currently do `impl ::enum_like::EnumLike` on the proc macro, should I just do `impl EnumLike` if I want to support reexporting the `EnumLike` trait under other names? What if someone does `extern crate enum_like as enl;` or something like that? My proc macro is forced to fail?

**Comment by alexcrichton** [comment]
Yes this definitely isn't great today unfortunately. The "real solution" is hygiene and until we get that a derive/macro basically needs to pick how it should be invoked and document that. For exampel @Badel2 you can use reexports yeah but you'd have to *only* do that effectively

**Comment by safinaskar** [comment]
This is duplicate of https://github.com/rust-lang/rust/issues/54363
